core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/handler/BaseAsyncClientHandler.java

```diff
@@ -60,44 +67,88 @@ protected BaseAsyncClientHandler(SdkClientConfiguration clientConfiguration,
-        ExecutionContext executionContext = createExecutionContext(executionParams);
 
-        HttpResponseHandler<OutputT> decoratedResponseHandlers =
-            decorateResponseHandlers(executionParams.getResponseHandler(), executionContext);
+        validateExecutionParams(executionParams);
+        ExecutionContext executionContext = createExecutionContext(executionParams, createInitialExecutionAttributes());
+        TransformingAsyncResponseHandler<Response<OutputT>> combinedResponseHandler;
+
+        if (executionParams.getCombinedResponseHandler() == null) {
+            // All other protocols take this path
+            HttpResponseHandler<OutputT> decoratedResponseHandlers =
+                decorateResponseHandlers(executionParams.getResponseHandler(), executionContext);
 
-        TransformingAsyncResponseHandler<OutputT> asyncResponseHandler =
-            new AsyncResponseHandler<>(decoratedResponseHandlers,
-                                       crc32Validator,
-                                       executionContext.executionAttributes());
+            TransformingAsyncResponseHandler<OutputT> successHandler =
+                new AsyncResponseHandler<>(decoratedResponseHandlers,
+                                           crc32Validator,
+                                           executionContext.executionAttributes());
+
+            TransformingAsyncResponseHandler<? extends SdkException> errorHandler =
+                resolveErrorResponseHandler(executionParams, executionContext, crc32Validator);
+            combinedResponseHandler = new CombinedResponseAsyncHttpResponseHandler<>(successHandler, errorHandler);
+        } else {
+            // XML non-streaming protocol takes this path
+            HttpResponseHandler<Response<OutputT>> decoratedResponseHandlers =
+                decorateSuccessResponseHandlers(executionParams.getCombinedResponseHandler(), executionContext);
+
+            combinedResponseHandler = new AsyncResponseHandler<>(decoratedResponseHandlers,
+                                                                 crc32Validator,
+                                                                 executionContext.executionAttributes());
+        }
 
-        return doExecute(executionParams, executionContext, asyncResponseHandler);
+        return doExecute(executionParams, executionContext, combinedResponseHandler);
     }
 
     @Override
     public <InputT extends SdkRequest, OutputT extends SdkResponse, ReturnT> CompletableFuture<ReturnT> execute(
         ClientExecutionParams<InputT, OutputT> executionParams,
         AsyncResponseTransformer<OutputT, ReturnT> asyncResponseTransformer) {
 
-        // For streaming requests, prepare() should be called as early as possible to avoid NPE in client
-        // See https://github.com/aws/aws-sdk-java-v2/issues/1268
+        validateExecutionParams(executionParams);
+
+        if (executionParams.getCombinedResponseHandler() != null) {
```

Comment:
should this also be moved to `validateExecutionParams`?

Reply:
That would break the non-streaming code-path as they both use validateExecutionParams and this additional check only applies to the streaming case.